j3soon/whisper-to-input
# Automatic Recording Cancellation / Finish & Enter Button Functionality

# What This Branch Accomplished

- **Speak Before I Get Bored**:
  - If the user does not speak for 2 seconds upon recording start, recording is automatically **cancelled**.
  - If the user stays rather silent for 1 second after speaking something, recording is automatically **finished** (thus starting transcribing).
  - These are achieved via `RecorderFSM` (see below).
- **Enter Enter**: An enter button replaces the "recording done" button. It is always visible unlike its predecessor.
  - If pressed during **Recording**, starts transcribing immediately. Also, appends a newline (`\r\n`) afterwards.
  - Otherwise, functions like a normal "Enter" button (used for newline, confirm input, start searching, etc).
  - Automatic recording finishes from **Speak Before I Get Bored** do not append newlines.

# Testing This Branch

Presented below are several recommended tests, but you can play around as you like.

1. Start recording and say nothing. The recording should terminate shortly (currently in 5 seconds).
2. Start recording, say something, and then stay silent. The transcribing should automatically start (currently in 3 seconds).
3. Start recording, say something, and press "Enter." Your text should be trailed with a newline.
4. Use the "Enter" normally (e.g., for typing newlines, or finishing a search etc.).

# Notes

- The `KeyEvent` code comes from #4. I forgot to mention it in my commit message.

# How `RecorderFSM` Works

## Finite State Machine

Amplitude reports should come in regularly, and for each report, `RecorderFSM` provides an action that the recorder should take (`Normal`, `CancelRecording`, and `FinishRecording`). The [mediator](https://en.wikipedia.org/wiki/Mediator_pattern) (currently `WhisperInputService`) will then notify the recorder of such actions.

`RecorderFSM` manages the recording behavior with 2 states: **Idle** and **Speaking**.
- **Idle**: The initial state when starting recording.
  - If any amplitude report exceeds `R.integer.recorder_fsm_idle_speaking_threshold`, state turns into **Speaking**.
  - If the user stays rather silent for at least `R.integer.recorder_fsm_idle_cancel_time` (ms), recording is cancelled, as if nothing happened. Specifically, all reports should be below `R.integer.recorder_fsm_idle_cancel_threshold` during this time.
- **Speaking**: The state in which the user is speaking.
  - The user can speak for an arbitrary amount of time.
  - If the user stays rather silent for at least `R.integer.recorder_fsm_speaking_finish_time` (ms), recording is finished, and transcription started. Specifically, all reports should be below `R.integer.recorder_fsm_speaking_finish_threshold` during this time.

## Accumulation-Based Algorithm

To check whether for the past `x` reports, all reports are valued below some threshold, we used a counter that increments when a report is below a threshold, and reset to 0 otherwise. 

When the counter becomes `x`, then for the past `x` reports, all reports are valued below that threshold.

## The Maximum in A Windowed Stream (No Longer Used)

**Note: This algorithm is no longer used. The MWS algorithm actually solves a harder problem than what is needed for `RecorderFSM`**.

The mechanism above involves **a stream of data** (amplitude reports), and detecting the maximum over the most recent some-**period-of-time** (window). (e.g., when the maximum report over the last 2 seconds is below some threshold, do something)

We refer to this problem as **"the maximum in a windowed stream"**.
To efficiently obtain the maximum in (or _over_) a windowed stream, several data is maintained.
1. **Data Window** `w`: Contains all the data points in the current window. As the stream feeds more data, older data of the window will be _popped_ from the front, and newer data will be appended to the end of the window.
2. **Decreasing Deque** `dq`: A deque whose element is a subset of the data window. More on this deque below.

Trivially, it is sufficient to just maintain the data window, if we want to obtain the maximum value in the window, but with time complexity **linear to window length** for every retrieval.

We would like to achieve amortized constant time here. It is not enough to just maintain the current maximum value, since if it is removed from the window, the (orig) second-max element in the window would now become the new max element.

Thus, we make use of a **decreasing deque** (`dq`), that has the following properties,
1. The elements in `dq` are a subset of those in `w`, with the same relative order. (e.g., `w=[2 5 1 3 4], dq = [5 4]`)
2. The first element of `dq` is _always_ the maximum value of the current window.
3. `dq` is a monotonically non-strict decreasing sequence (aka. a non-increasing sequence).

If `dq` is properly maintained, then we can obtain the current maximum in constant time. The current maximum can only change when a new value arrives. When it happens,
1. If the current window is full, an old value should be popped first. And if the popped value is the **current maximum**, a value should be popped from `dq` as well.
2. Then, the arriving value is appended to the back of `w`.
3. This new value should be appended to the back of `dq`. However, to maintain the properties of `dq`, elements will be popped from the back until the new value is the minimum in `dq` (decreasing property). For example, if the new value is extremely large, all values originally in `dq` will be popped. The new value now becomes the head of `dq`, also the current maximum.

Below is a simple simulation of this algorithm implemented in Python,

```python
import random

WINDOW_LEN = 10
window = []
deque = []

for _ in range(1_000_000):
    new_val = random.randint(0, 150)
    
    if len(window) == WINDOW_LEN:
        if deque[0] == window[0]:
            deque = deque[1:]
        window = window[1:]

    window.append(new_val)

    while len(deque) > 0 and deque[-1] < new_val:
        deque = deque[:-1]
    deque.append(new_val)

    true_max = max(window)
    assert(true_max == deque[0])
```

---

Depends on #18
